cmd/juju/application/refresh.go

```diff
@@ -502,7 +502,7 @@ func (c *refreshCommand) isCharmHubWithRevision(source commoncharm.OriginSource)
 		return true
 	}
 	// EnsureSchema will error if input is an empty string.
-	path, err := charm.EnsureSchema(c.SwitchURL)
+	path, err := charm.EnsureSchema(c.SwitchURL, charm.CharmHub)
```

Comment:
Why are we assuming the switch charm must be charmhub if not specified, with a 2.8 controller and 2.9 client?

$ juju refresh ceph-mon --switch influxdb
Looking up metadata for charm-store charm "ceph-mon"
ERROR charmhub charms are not supported by the current controller; if you wish to use charmhub consider upgrading your controller to 2.9+.
$ juju refresh ceph-mon --switch cs:influxdb
Looking up metadata for charm-store charm "ceph-mon"
Added charm-store charm "influxdb", revision 23 in channel stable, to the model
Adding endpoint "grafana-source" to default space "alpha"
Adding endpoint "query" to default space "alpha"
Leaving endpoints in "alpha": admin, nrpe-external-master

Reply:
We shouldn't. I was under the wrong impression that this codepath was only used for 2.9+. I will update it to use the same logic as the deployer.

Reply:
Pushed https://github.com/juju/juju/pull/12948/commits/1ce13dfedb0124b24550cc2bd865bf61fa26be17 to address these comments and updated QA steps.